src/electos/ballotmaker/data/extractor.py

```diff
@@ -0,0 +1,311 @@
+    CandidateSelection,
+    Election,
+    ElectionReport,
+    InternationalizedText,
+    OrderedContest,
+    OrderedHeader,
+)
+
+
+# --- Base Types
+#
+# Schema expresses these as union types not subclasses
+
+Contest = Union[BallotMeasureContest, CandidateContest]
+
+OrderedContent = Union[OrderedContest, OrderedHeader]
+
+
+# --- Utilities
+
+
+def _text_content(item):
+    """Return joined lines from internationalized text."""
+    assert isinstance(item, InternationalizedText)
+    text = "\n".join(_.content for _ in item.text)
+    return text
+
+
+def _walk_ordered_contests(content: List[OrderedContent]):
+    """Walk ordered content yielding contests."""
+    for item in content:
+        if isinstance(item, OrderedContest):
+            yield item
+        elif isinstance(item, OrderedHeader):
+            yield from _walk_ordered_contests(item.ordered_content)
+        else:
+            raise TypeError(f"Unexpected type: {type(item).__name__}")
+
+
+def _walk_ordered_headers(content: List[OrderedContent]):
+    """Walk ordered content yielding headers."""
+    for item in content:
+        if isinstance(item, OrderedHeader):
+            yield item
+            yield from _walk_ordered_headers(item.ordered_content)
+        else:
+            raise TypeError(f"Unexpected type: {type(item).__name__}")
+
+
+# --- Extractor
+
+class BallotDataExtractor:
+
+    """Extract election data from an EDF."""
+
+
+    def __init__(self):
+        pass
+
+
+    def _ballot_style_external_id(self, ballot_style: BallotStyle):
+        """Get the text of a ballot style's external identifier if any."""
+        if ballot_style.external_identifier:
+            assert len(ballot_style.external_identifier) == 1, \
+                "Not ready to handle multiple BallotStyle external IDs"
+            name = ballot_style.external_identifier[0].value
+        else:
+            name = ""
+        return name
+
+
+    def _ballot_style_gp_units(self, ballot_style: BallotStyle):
+        """Yield geo-political units for a ballot style."""
+        for id_ in ballot_style.gp_unit_ids:
+            gp_unit = self._index.by_id(id_)
+            yield gp_unit
+
+
+    def _ballot_style_contests(self, ballot_style: BallotStyle):
+        """Yield the contests of a ballot style."""
+        for item in _walk_ordered_contests(ballot_style.ordered_content):
+            contest = self._index.by_id(item.contest_id)
+            yield contest
+
+
+    def _candidate_name(self, candidate: Candidate):
+        """Get the name of a candidate as it appears on a ballot."""
+        name = _text_content(candidate.ballot_name)
+        return name
+
+
+    def _candidate_party(self, candidate: Candidate):
+        """Get the name and abbreviation of the party of a candidate as it appears on a ballot.
+
+        Drop either field from result if it isn't present.
+        """
+        # Note: party ID is returned to allow de-duplicating parties in callers.
```

Comment:
The party ID being returned is a hack. It allows the caller to disambiguate the complication of multiple parties being shared by all candidates in a slate. I don't want to return an ID: that obliges the caller to have an index to look the party up. (It does, but that's an artifact of this particular code), and it would make this function behave oddly anyway, since every other function is returning a dictionary. If it's going to be a hack anyway it seems less leaky to me to have the data and ignore the ID.

The comment isn't a docstring because the ID being returned is an implementation detail, and I usually avoid implementation details in docstrings. Admittedly this is an internal function anyway, so it's all implementation, so I'll move it in to the docstring in this case.